jgrapht-core/src/test/java/org/jgrapht/alg/tour/TwoApproxMetricTSPTest.java

```diff
@@ -125,7 +128,29 @@ public void testStar()
         double mstWeight = new KruskalMinimumSpanningTree<>(g).getSpanningTree().getWeight();
         double tourWeight = tour.getWeight();
         assertTrue(2 * mstWeight >= tourWeight);
+    }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void testInvalidInstanceDirected()
+    {
+        DirectedGraph<String, DefaultEdge> g = new SimpleDirectedGraph<>(DefaultEdge.class);
+        g.addVertex("A");
+
+        new TwoApproxMetricTSP<String, DefaultEdge>().getTour(g);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testInvalidInstanceNotComplete()
+    {
+        SimpleWeightedGraph<String, DefaultWeightedEdge> g =
+            new SimpleWeightedGraph<>(DefaultWeightedEdge.class);
+        g.addVertex("A");
+        g.addVertex("B");
+        g.addVertex("C");
+        g.setEdgeWeight(g.addEdge("A", "B"), 20d);
+        g.setEdgeWeight(g.addEdge("A", "C"), 42d);
+
+        new TwoApproxMetricTSP<String, DefaultWeightedEdge>().getTour(g);
     }
 
     private static <V, E> void assertHamiltonian(UndirectedGraph<V, E> g, GraphPath<V, E> path)
```

Comment:
sorry to get back to this. I would propose to extend this method with some more checks. Most likely, this particular method will be reused by future TSP algorithms. Here's a (not tested) proposal:

```
private static <V, E> void assertHamiltonian(UndirectedGraph<V, E> g, GraphPath<V, E> path)
{
  List<V> tourVertices = path.getVertexList();
  List<E> tourEdges = path.getEdgeList();

  //check tour length, beginning and end of the tour
  assertEquals(path.getStartVertex(), path.getEndVertex());
  assertEquals(path.getStartVertex(), tourVertices.get(0));
  if(g.vertexSet().size()==1){
    assertTrue(tourEdges.isEmpty());
    assertEquals(path.getWeight(), 0.0);
    return;
  }
  assertEquals(g.vertexSet().size(), tourVertices.size()-1);
  assertEquals(g.vertexSet().size(), tourEdges.size());
  assertEquals(tourVertices.get(0), tourVertices.get(tourVertices.size()-1));

  //check that the tours identified by the vertex list and edge lists are identical
  //validate the weight of the tour as well;
  double weight=0;
  int index=1;
  V v = path.getStartVertex();
  for (E e : edges) {
      v = Graphs.getOppositeVertex(g, e, v);
      assertEquals(tourVertices.get(index), v);
      weight+=g.getEdgeWeight(e);
      index++;
  }
  assertEquals(path.getWeight(), weight);

  //check that every city is visited
  Set<V> unique=new HashSet<>(tourVertices);
  assertEquals(g.vertexSet().size(), tourVertices.size());
}
```